### PR TITLE
fix: fix typo and add some tests

### PR DESCRIPTION
**Pull request by hezhizhen**
I also tweaked ErrorCode usage and correct some format cases.

### CODE REVIEW

**Comment by hezhizhen** [comment]
```go

func (mb *MemDatabase) Delete(namespace string, key string) *DbError {
	mb.mu.Lock()
	defer mb.mu.Unlock()

	ns, ok := mb.namespaces[namespace]
	if !ok {
		return &DbError{
			ErrorCode: NAMESPACE_NOT_FOUND,
			Message:   fmt.Sprintf("namespace '%v' does not exist.", namespace),
		}
	}
	_, ok = ns.data[key]
	if !ok {
		return &DbError{
			ErrorCode: ID_NOT_FOUND,
			Message:   fmt.Sprintf("value not found in namespace '%v' for key '%v'", namespace, key),
		}
	}

	delete(ns.data, key)
	return nil
}
```

Btw, I think the second check is not necessary.

**Comment by rehacktive** [comment]
hello - thanks for your PR :) could you make sure all tests are passed?

**Comment by hezhizhen** [comment]
> hello - thanks for your PR :) could you make sure all tests are passed?

Sorry, I didn't notice it. I'll address it later.

**Comment by hezhizhen** [comment]
@rehacktive Would you run workflows again?

**Comment by hezhizhen** [comment]
According to the test report for the last PR https://github.com/rehacktive/caffeine/runs/4444796113?check_suite_focus=true , there are some tests failing to pass, but they are ignored.

**Comment by hezhizhen** [comment]
> Hello,
> 
> I'm the commiter who put in the error checking.
> 
> You mention above that you don't think this check is necessary:
> 
> ```
> 	_, ok = ns.data[key]
> 	if !ok {
> 		return &DbError{
> 			ErrorCode: ID_NOT_FOUND,
> 			Message:   fmt.Sprintf("value not found in namespace '%v' for key '%v'", namespace, key),
> 		}
> 	}
> ```
> 
> I would suggest that it is useful, because it helps to return a more informative error message in case a key is not found in the database.
> 
> Basically it allows the API to send back a 404 for items that are not found. Without this additional error detail, it wouldn't be possible to distinguish 'unexpected errors' (which generate 500 errors) and cases where the key is simply not found.
> 
> Also, beware that changing these error responses will also change the behaviour of the API. If you remove this code and someone downstream of the caffeine API server expects to see 404s, they will never receive them anymore.
> 
> Sorry in advance if I misunderstood your comment above.

Hi, thank you for your explanation, and I fully understand that. Here's my opinion from another aspect.

When we perform a deletion operation, the ultimate status is that the resource is gone. Now that the key doesn't exist, which means we already reach the status we expect, why do we need to return an error to say that we are performing an invalid operation (delete a resource that doesn't exist)?

> If you remove this code and someone downstream of the caffeine API server expects to see 404s, they will never receive them anymore.

It's true, and that's one of the reasons that I think this check should be kept.

Btw, I wonder if there're some guidelines about when to choose which one.

**Comment by hezhizhen** [comment]
@rehacktive I don't know why the tests are not passed, so I imported go-cmp to see the diff. Would you run workflows again?

**Comment by hezhizhen** [comment]
> 

All I need is a new blank line... Anyway, I replace it with an existing function to retrieve the JSON from the JSON file.

**Comment by rehacktive** [comment]
I think the 404 approach would make all cleaner, and no cons. agreed?  @hezhizhen  @annoybot 

test passed, gonna check something and merge, thanks both :)

File: database/mem_database_test.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package database
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestMemDatabase_Upsert(t *testing.T) {
```
Comment (rehacktive) [inline database/mem_database_test.go:8]: what's the point on adding this? there's already a list of tests running on different implementations, including the mem database.

Comment (hezhizhen) [reply database/mem_database_test.go:8]:
```
func (mb *MemDatabase) Upsert(namespace string, key string, value []byte) *DbError {
	mb.mu.Lock()
	defer mb.mu.Unlock()
	ns, ok := mb.namespaces[namespace]
	if !ok {
		ns = newNamespace()
		mb.namespaces[namespace] = ns
	}
	ns.data[key] = value
	return nil
}
```

I was not sure if the upsert function worked as expected, so I added this test. I thought `mb.namespaces[namespace]=ns` was necessary, but it turned out not.

**Comment by hezhizhen** [comment]
> Hi, thanks for your response.
> 
> > When we perform a deletion operation, the ultimate status is that the resource is gone. Now that the key doesn't exist, which means we already reach the status we expect, why do we need to return an error to say that we are performing an invalid operation (delete a resource that doesn't exist)?
> 
> I'll make a few arguments:
> 
> **Multiple clients:** Say you have two clients connected to this API server and that there exists a JSON under the key `42`. Now client one deletes the key `42`. Later client two comes looking for key `42`, but has no knowledge that client 1 deleted it. A 404 would allow client two to realise that this key does not exist.
> 
> If we were to return a 400, it would leave the client wondering, what's wrong with my request, how should I rewrite it in order to find `42`? 404 means, your request is well formed, but there is no key `42`. 400 (which this server used to return for missing keys) means your request is malformed.
> 
> **Precedent:** If you setup a HTTP server like apache or NGINX. You put a picture of a cat on there. Some people download it and pass the URL around to their friends (the cat is _really_ cute and doing something funny). But one day you delete this picture from NGINX. Now when people request that URL, they get a 404. Basically, it's customary in HTTP to get a 404 for a resource that is not found. Web browsers are stateless, i.e.: they typically do not rely on their knowledge of previous requests to guide their decisions. HTTP was designed with statelessness in mind.
> 
> **More detail is better:** The more detail you preserve in error codes, the better. In this case the database layer is preserving enough information about the kind of error that happened to let other higher layers in the code make fully informed decisions on how to handle it. In this case, the error detail is preserved through the next layer, where we send either a 404 or a 400 on down to the caller. This allows the caller to decide how to handle it. To one user of this API server it could be a fatal error to encounter a missing key, to another it could be something less serious.

I think your examples are not quite appropriate. When a client deletes a key, another client should and will get a 404 error when it looks for the same key. But when it tries to delete the same key, should it get a 404 error or just be told it's already done? That's the question we are arguing about. But I agree with the third point, and that's what @rehacktive concluded later.

**Comment by hezhizhen** [comment]
> > I think your examples are not quite appropriate. When a client deletes a key, another client should and will get a 404 error when it looks for the same key. But when it tries to delete the same key, should it get a 404 error or just be told it's already done?
> 
> That's just it the second client doesn't need to be told that the deletion is 'already done'. The second client doesn't need to know that, all it needs to know is that there is no item with that id to be deleted.
> 
> Look at the RFC: https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.1 It specifies that 400 implies that the request
> 
> > The 400 (Bad Request) status code indicates that the server cannot or
> > will not process the request due to something that is perceived to be
> > a client error (e.g., malformed request syntax, invalid request
> > message framing, or deceptive request routing).
> 
> Because of the statelessness of HTTP and the possibility of several clients or sessions, I think it's still better for a delete to send back a 404 if the key isn't found.
> 
> Requesting to delete a non-existent key isn't an error with the request. It only means the called didn't know if the key existed or not.
> 
> Put it this way, if we send back 400 for deleting a non-existent key, then the only way for the client to distinguish that condition from other 'malformed request' errors, would be for the client to first call GET for that id, and deal with the 404 returned from that call instead. So now we'd need two calls instead of one.
> 
> This just seems unnecessary to me.

> Put it this way, if we send back 400 for deleting a non-existent key...

I guess you misunderstand what I said. Client A deletes key `42` and it gets an OK status. Client B doesn't know that the key has already been deleted and it deletes it as well. I think it should receive an OK status instead of 404 or 400, which means key `42` is deleted (although it is NOT deleted by client B).

**Comment by hezhizhen** [comment]
> Ah yes I see what you mean.
> 
> You know there's an expression: "There's no discussing matters of taste," which is what I feel this discussion is becoming.
> 
> Some APIs always return a 200 for DELETE, some always return a 204. Others will return 404 for missing keys.
> 
> Personally I prefer the most detail possible. There's really no downside for the client code. If the API were designed to always return 200 the code would check for success like this: `if http_response_code == 200 then request succeeded... endif`.
> 
> If it could return a 200 or a 404 and the client doesn't care about the 404s, it can just handle success like this `if http_response_code == 200 || http_response_code == 404 then request succeeded... endif`
> 
> I think more detail is better, because we have no idea of the needs of all the clients which could make use of this API server. So it's best to give them all the information they need to make their own decisions.

I totally agree with it.

**Pull request merged**